src/Ledger/Gov.lagda

```diff
@@ -41,6 +51,57 @@ _⊢_⇀⦇_,GOV⦈_     : GovEnv → GovState → List (GovVote ⊎ GovProposal
 \begin{code}[hide]
 open GovActionState
 
+-- convert list of (GovActionID,GovActionState)-pairs to list GovActionID pairs.
+getAidPairsList : GovState → List (GovActionID × GovActionID)
+getAidPairsList aid×states =
+  mapMaybe (λ (aid , aState) → (aid ,_) <$> getHash (prevAction aState)) $ aid×states
+
+-- convert list of (GovActionID,GovActionState)-pairs to SET of GovActionID pairs.
+getAidPairsSet : GovState → ℙ (GovActionID × GovActionID)
+getAidPairsSet aid×states =
+  mapPartial (λ (aid , as) → (aid ,_) <$> getHash (prevAction as)) $ fromList aid×states
+
+-- a list of GovActionID pairs connects the first GovActionID to the second
+_connects_to_ : List (GovActionID × GovActionID) → GovActionID → GovActionID → Set
+[] connects aidNew to aidOld = aidNew ≡ aidOld
+((aid , aidPrev) ∷ s) connects aidNew to aidOld = aid ≡ aidNew × s connects aidPrev to aidOld
+                                                  ⊎ s connects aidNew to aidOld
+-- TODO: delete these notes (and possibly add them to docs) after confirming they're correct.
+--   Suppose:  aid₁ → aid₂ → aid₃ → aid₄ → aid₅.
+--   And let:  l := (aid₅, aid₄) (aid₄, aid₃) (aid₃, aid₂) (aid₂, aid₁)
+--   Check:    l connects aid₅ to aid₂ :
+--               ((aid₅, aid₄) ∷ s) connects aid₅ to aid₂ = aid₅ ≡ aid₅ × s connects aid₄ to aid₂
+
+[_connects_to_?] : ∀ l aidNew aidOld → Dec (l connects aidNew to aidOld)
+[ [] connects aidNew to aidOld ?] = aidNew ≟ aidOld
+
+[ (aid , aidPrev) ∷ s connects aidNew to aidOld ?] =
+  ((aid ≟ aidNew) ×-dec [ s connects aidPrev to aidOld ?]) ⊎-dec [ s connects aidNew to aidOld ?]
+
+enactable : EnactState → List (GovActionID × GovActionID) → GovActionID × GovActionState → Set
+enactable e aidPairs = λ (aidNew , as) → case getHashES e (GovActionState.action as) of λ where
+  nothing → ⊤
+  (just aidOld) → ∃[ t ](fromList t ⊆ fromList aidPairs × t connects aidNew to aidOld)
+
+open Equivalence
+
+∃?-connecting-subset : ∀ L {u}{v} → Dec (∃[ l ](l ⊆ˡ L) × (l connects u to v))
+∃?-connecting-subset L {u}{v} with [ L connects u to v ?]
+... | yes p = yes (L , (⊆ˡ-id , p))
+... | no ¬q = {!!}
```

Comment:
This is the last (and hardest) hole that's left to fill.  I believe a recursive algorithm would work---one that looks at each subpermutation of L of length |L|-1 to find one that connects u to v.

Reply:
The whole point of adding the proofs about subpermutations is to make this easy, since `∃[ l ](l ⊆ˡ L) × (l connects u to v)` is equivalent to `Any (_connects u to v) (subpermutations L)` and decidability of that is just `any? [_connects u to v ?] (subpermutations L)`.

Reply:
Yes, I know that, but I wasn't able to prove the equivalence `[ l ](l ⊆ˡ L) × (l connects u to v) ⇔ Any (_connects u to v) (subpermutations L)` so I was trying to find an alternative route... but I'll give the equivalence proof another try today.

Reply:
For example, one of the issues is that `[] ⊆ˡ []` but `[] ∉ subpermutations []` (since `subpermutations []` is `[]` and `[] ∉ []`).  Of course, `[]` will not be a "connecting list" (unless the two action ids in question are equal), so I should be able to find a way around this issue... still working on it.

Reply:
I finally managed to prove the equivalence `l ⊆ˡ L ⇔ Subperm l L` for the inductive `Subperm` type I had defined earlier, so now I just need to get the "Any" function working for that type and I think we're in business!

Reply:
I thought the whole point of `subpermutations` was that it'd be a function that returns all the sublists. If it doesn't do that then you should probably fix that, otherwise `subpermutations` will just be completely unused. And note that the `subpermutations` function is the one doing all the work, so replacing it means you have to write another function constructing subpermutations.

Also, this reveals that you've had a bug in your changed definition: if `subpermutations [] = []`, then if the list of proposals is empty, `enactable` is always the empty type!

Reply:
Yes, I agree with all of that.  Actually, `subpermutations` is fine.  I've tested it.  It gives the expected results on small lists, empty list, etc.  However (as you observed early on) it's hard to prove that it is indeed the list of all sublists, and permutations thereof, of the given list.

I agree that if I could prove that `(l ⊆ˡ L) ⇔ l ∈ subpermutations L`, then I could use `any?` on the list `subpermutations aidPairs` to complete the decidability proof (`∃?-connecting-subset`)... and I'll be done!

Thanks for your comments and helping me clarify what I need to do. 👍🏼

Reply:
Update: I think proving `Subperm l L ⇔ l ∈ subpermutations L` may be easier.  I just pushed an outline of the proof to `Data.List.Ext.Subperm.Properties`.  I'll try to complete the proof today.

Reply:
I'm really confused why you're still going ahead with this. The theorem you eventually need to prove is
```
l ⊆ L ⇔ l ∈ subpermutations L
```
but your `subpermutations` function still doesn't have that property. You seem to be proving that equivalence in the case that `l` isn't the empty list, but why? The whole point of the `subpermutations` function is to be a function that gives you a list having exactly this property. Just change `subpermutations` to have the property you need!

Most likely this will also make a bunch of proofs substantially easier. You can even write it in a correct-by-construction manner, where you first define `subpermutationsᶜ : (L : List A) → List (∃[ l ] Subperm l L)` and then `subpermutations` just projects out the lists. Then the first half of the equivalence is free.

Reply:
I don't entirely agree with all of the points above.  In particular, the property `l ⊆ L ⇔ l ∈ subpermutations L` will not (and should not) hold in general.  It will hold for lists without repeated entries... and I am indeed trying to prove the equivalence under those assumptions. I hope to complete the proof today.  I am very close.